### ISSUE TITLE
Discriminator models with nested schemas broke between 3.8.16 and 4.0.1

### ISSUE THREAD

**Issue by stephenbunch** [Automattic/mongoose#2821]
Hey guys, I'm not sure what changed, but the following code breaks in 4.0.1:

```js
var mongoose = require( 'mongoose' );
var util = require( 'util' );

function MinionSchema() {
  mongoose.Schema.apply( this, arguments );

  this.add({
    name: String
  });
}
util.inherits( MinionSchema, mongoose.Schema );

function BaseSchema() {
  mongoose.Schema.apply( this, arguments );

  this.add({
    name: String,
    created_at: Date,
    minions: [ new MinionSchema() ]
  });
}
util.inherits( BaseSchema, mongoose.Schema );

var PersonSchema = new BaseSchema();
var BossSchema = new BaseSchema({
  department: String
});

var Person = mongoose.model( 'Person', PersonSchema );
var Boss = Person.discriminator( 'Boss', BossSchema );
```

```
TypeError: Object #<Object> has no method '_markModified'
    at EmbeddedDocument.markModified (/Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/types/embedded.js:63:24)
    at EmbeddedDocument.Document.$__set (/Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/document.js:666:10)
    at EmbeddedDocument.Document.set (/Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/document.js:604:10)
    at EmbeddedDocument.Object.defineProperty.set [as name] (/Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/document.js:1418:45)
    at merge (/Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/utils.js:362:15)
    at merge (/Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/utils.js:364:7)
    at merge (/Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/utils.js:364:7)
    at Object.merge (/Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/utils.js:364:7)
    at /Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/model.js:724:11
    at Function.discriminator (/Users/stephen/code/pinwheels/pinwheels/node_modules/mongoose/lib/model.js:759:5)
```

The error doesn't occur unless I define an attribute on the MinionSchema.

So I stepped through the `utils.merge` function and noticed that when it gets to merging the `minion` path in 3.8.16, the caster is an EmbeddedDocument with its prototype set to a Minion document. In 4.0.1, the caster itself is a Minion document. The result is that when Object.keys is called, the Minion properties are included and the merge function triggers the Minion setter logic.

But I can't make heads or tails of this. The constructor of an EmbeddedDocument expects an `Object` and a `MongooseDocumentArray`, but `SchemaArray` initializes a new EmbeddedDocument passing `null` and `castOptions`, and it appears it's been this way for a while. I'm super confused.

**Comment by vkarpov15** [comment]
Thanks for pointing this out, I'll investigate later

**Comment by vkarpov15** [comment]
I took a look and I can confirm that this is a bug, but I don't have a fix atm. Works in 3.8.25 but not in 4.0.1. I'll take a closer look early next week, although any help would be much appreciated.

**Comment by jkanschik** [comment]
Additional information: works in 3.9.0, but not in 3.9.1, maybe this info helps a bit.

**Comment by vkarpov15** [comment]
Bisected this down to 6ce5a6c0d409fe2bb3a2f36bc13716e55467640d and put in a PR #2863 to fix. Admittedly that commit was not my best work - thanks for finding this issue.

**Comment by mtgibbs** [comment]
This fix finding its way back into 3.8.x?

**Comment by vkarpov15** [comment]
@mtgibbs does this affect 3.8.x? AFAIK it shouldn't.

**Comment by mtgibbs** [comment]
@vkarpov15 I walked back to the 3.8.x package while waiting for the 4.0.2 package and I see the same symptoms.  I'm pretty new to using mongoose, so I could just being messing stuff up between flipping back and forth between the versions and how I have my schema defined. I plan to test it out with the pending 4.0.2 fix you applied in 3.8.x tonight and confirm it.

**Comment by mtgibbs** [comment]
@vkarpov15 I finally checked the file in my node_modules for mongoose and I was pulling a 3.9.x version from npm that I didn't realize was available because I defined the dependency 3.8.x with a caret.... /bonk.  The test that stephen provided works just fine in 3.8.25.  Thanks for the help.

**Comment by vkarpov15** [comment]
Please don't use `^` with mongoose 3.x - mongoose 3 used [linux-style stable/unstable versioning](https://github.com/Automattic/mongoose/wiki/Development-%28unstable%29-branch-details) which is **not** compatible with semver. Mongoose 4 [does use semver](https://github.com/Automattic/mongoose/wiki/4.0-Release-Notes). Use "~3.8" to target latest 3.8.x.
